lib/bolt/transport/docker/connection.rb

```diff
@@ -8,60 +8,85 @@ module Transport
         end
 
         def connect
-          # Explicitly create the new Connection to avoid relying on global state in the Docker module.
-          url = @target.options['service-url'] || ::Docker.url
-          options = ::Docker.options.merge(@target.options['service-options'] || {})
-          @container = ::Docker::Container.get(@target.host, {}, ::Docker::Connection.new(url, options))
+          # We don't actually have a connection, but we do need to
+          # check that the container exists and is running.
+          output = execute_local_docker_json_command('ps')
+          index = output.find_index { |item| item["ID"] == @target.host || item["Names"] == @target.host }
+          raise "Could not find a container with name or ID matching '#{@target.host}'" if index.nil?
+          # Now find the indepth container information
+          output = execute_local_docker_json_command('inspect', [output[index]["ID"]])
+          # Store the container information for later
+          @container_info = output[0]
           @logger.debug { "Opened session" }
+          true
         rescue StandardError => e
           raise Bolt::Node::ConnectError.new(
             "Failed to connect to #{@target.uri}: #{e.message}",
             'CONNECT_ERROR'
           )
         end
 
+        # Executes a command inside the target container
+        #
+        # @param command [Array] The command to run, expressed as an array of strings
+        # @param options [Hash] command specific options
+        # @option opts [String] :interpreter statements that are prefixed to the command e.g `/bin/bash` or `cmd.exe /c`
+        # @option opts [Hash] :environment A hash of environment variables that will be injected into the command
+        # @option opts [IO] :stdin An IO object that will be used to redirect STDIN for the docker command
         def execute(*command, options)
           command.unshift(options[:interpreter]) if options[:interpreter]
+          # Build the `--env` parameters
+          envs = []
           if options[:environment]
-            envs = options[:environment].map { |env, val| "#{env}=#{val}" }
-            command = ['env'] + envs + command
+            options[:environment].each { |env, val| envs.concat(['--env', "#{env}=#{val}"]) }
           end
 
-          @logger.debug { "Executing: #{command}" }
-          result = @container.exec(command, options) { |stream, chunk| @logger.debug("#{stream}: #{chunk}") }
+          command_options = []
+          # Need to be interactive if redirecting STDIN
+          command_options << '--interactive' unless options[:stdin].nil?
+          command_options.concat(envs) unless envs.empty?
+          command_options << container_id
+          command_options.concat(command)
+
+          @logger.debug { "Executing: exec #{command_options}" }
+
+          result = execute_local_docker_command('exec', command_options, options[:stdin])
```

Comment:
Would be better to use destructuring assignment here for readability like is done below:

```ruby
stdout_str, stderr_str, status = execute_local_docker_command('exec', command_options, options[:stdin])
```

Reply:
That'll teach for me copy-pasta the original code :-).  Will do.

Reply:
Done.